### PR TITLE
31 improve express middleware

### PR DESCRIPTION
**Pull request by ISO50**

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -99,10 +99,37 @@ export function extractAccessToken(authHeader: string): string {
...
+      grant_type,
+      authorization_code,
+      cn,
+      realm,
+      token_type,
+      expires_in,
+      client_id
+    } = data;
+
+    const tokeninfo = {
```
Comment (bzums) [inline src/utils.ts]: remove `grant_type`, `authorization_code`, `realm` and `client_id` and `token_type`

File: src/express-tooling.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+function requireScopesMiddleware(scopes: string[], precedenceFunction?: Function) {
+  return function(req: any, res: any, next: Function) {
+
+    if (precedenceFunction) {
+      precedenceFunction(req, res, next)
+      .then(result => {
+        if (result) {
+          next();
+        }
+      });
```
Comment (bzums) [inline src/express-tooling.ts]: `return`

File: src/express-tooling.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ * Usage:
+ *  app.get('/path', requireScopesMiddleware['scopeA', 'scopeB'], (req, res) => { // Do route work })
+ *
+ * @param scopes - array of scopes that are needed to access the endpoint
+ * @param precedenceFunction - must return a promise that return true or false.
+ *                             If the result is true the scope checking will be skipped
+ *                             and next is called
+ * @returns {function(any, any, any): undefined}
+ */
+function requireScopesMiddleware(scopes: string[], precedenceFunction?: Function) {
```
Comment (bzums) [inline src/express-tooling.ts]: add function signature?
`precedenceFunction?:  (...args: any[]): Promise<boolean>

File: src/utils.ts [hunk 4]
```diff
@@ -99,10 +99,29 @@ export function extractAccessToken(authHeader: string): string {
...
+export function setTokeninfo(req: express.Request) {
   return function(data: any) {
+
+    const {
+      uid,
+      scope,
+      cn,
+      /* tslint:disable */
+      expires_in,
+      /* tslint:enable */
```
Comment (bzums) [inline src/utils.ts]: `// tslint:disable-line` should wor as well

File: src/express-tooling.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ */
+function requireScopesMiddleware(scopes: string[],
+                                 precedenceFunction?: IPrecedenceFunction) {
+
+  return function(req: any, res: any, next: Function) {
+
+    if (precedenceFunction) {
+      precedenceFunction(req, res, next)
+      .then(result => {
+        if (result) {
```
Comment (bzums) [inline src/express-tooling.ts]: 2 would most probably mean that we need one or two additional test I guess

**Comment by bzums** [comment]
👍

**Comment by ISO50** [comment]
👍

**Pull request merged**